### PR TITLE
Feature/324 implement update locations

### PR DESCRIPTION
**Pull request by brendan-oconnell**
Addresses https://github.com/thoth-pub/thoth/issues/324

Known bug: When a user views a Canonical Location and exits the modal without making changes by clicking Cancel, X, or outside of the modal, and then opens the modal for another non-Canonical Location, the second Location will sometimes incorrectly display as Canonical: Yes in the modal. 

If the user then returns to the Canonical Location without making changes to the second Location, the Canonical Location will sometimes incorrectly display Location Platform: Project MUSE. 

In neither case does this affect the underlying values of the fields upon Save; if the user clicks Save when the values are displaying incorrectly, the correct, original values of the fields will be preserved. 

This display issue has already been identified in several other parts of the view (see https://github.com/thoth-pub/thoth/issues/366).

### CODE REVIEW

File: thoth-app/src/models/location/update_location_mutation.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    ) {
+        updateLocation(data: {
+            locationId: $locationId
+            publicationId: $publicationId
+            landingPage: $landingPage
+            fullTextUrl: $fullTextUrl
+            locationPlatform: $locationPlatform
+            canonical: $canonical
+        }){
+            locationId
```
Comment (rhigman) [inline thoth-app/src/models/location/update_location_mutation.rs:24]: Not all fields need to be retrieved here - e.g. I don't think `createdAt` or `updatedAt` are used in the display

File: thoth-app/src/models/location/location_query.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
+use serde::Deserialize;
```
Comment (rhigman) [inline thoth-app/src/models/location/location_query.rs:1]: I can't see where this module is used.

File: thoth-app/src/models/mod.rs [hunk 3]
```diff
@@ -17,8 +17,9 @@ macro_rules! graphql_query_builder {
         use yewtil::fetch::MethodBody;
 
         use $crate::THOTH_GRAPHQL_API;
-
+        #[allow(dead_code)]
```
Comment (rhigman) [inline thoth-app/src/models/mod.rs:20]: Why are these needed?

**Review by rhigman — COMMENTED** [review]
Haven't tried running the code yet, just read through it and noted a few things that could be tweaked.

File: CHANGELOG.md [hunk 4]
```diff
@@ -16,6 +16,910 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
   - [565](https://github.com/thoth-pub/thoth/issues/565) - Don't generate Crossref metadata output if no DOIs (work or chapter) are present
+  - [324](https://github.com/thoth-pub/thoth/issues/324) - Make Locations editable, including the ability to change the Canonical Location for a Publication.
+
+## [[0.12.9]](https://github.com/thoth-pub/thoth/releases/tag/v0.12.9) - 2024-09-06
```
[the remark this answers is not in the document]
Comment (brendan-oconnell) [reply CHANGELOG.md]: Oops, thanks - I should have caught that myself!

File: thoth-api/src/graphql/model.rs [hunk 5]
```diff
@@ -1984,21 +1989,74 @@ impl MutationRoot {
...
+                    .first::<Location>(&mut context.db.get()?);
+
+                let final_canonical_location = match canonical_location {
+                    Ok(location) => location,
+                    Err(e) => {
+                        return Err(ThothError::from(e).into());
+                    }
+                };
+
+                let old_canonical_location = Some(PatchLocation {
```
[the remark this answers is not in the document]
Comment (brendan-oconnell) [reply thoth-api/src/graphql/model.rs]: Good question, I think this was an artifact of getting things to work much earlier in the process. I've removed them.

File: thoth-api/src/graphql/model.rs [hunk 6]
```diff
@@ -1984,21 +1989,74 @@ impl MutationRoot {
...
+                    location_id: final_canonical_location.location_id,
```
[the remark this answers is not in the document]
Comment (brendan-oconnell) [reply thoth-api/src/graphql/model.rs]: `final_canonical_location` is a `Location` object, whereas `old_canonical_location` needs to be a `PatchLocation` object when it gets passed to `update_canonical_location()`. So simply making a copy of `final_canonical_location` won't work, but if there's a more efficient way to perform the operation I need, I'd appreciate any ideas!

File: thoth-api/src/graphql/model.rs [hunk 7]
```diff
@@ -1984,21 +1989,74 @@ impl MutationRoot {
...
+
+                        // Update the current non-canonical location to canonical
+                        diesel::update(location::table.find(data.location_id))
+                            .set(&data)
+                            .execute(connection)?;
+
+                        Ok::<_, diesel::result::Error>(())
+                    });
+                }
+                Ok(location)
```
[the remark this answers is not in the document]
Comment (brendan-oconnell) [reply thoth-api/src/graphql/model.rs]: I've refactored this; see `update_canonical_location()` in `location/crud.rs`. Now I have it just returning `Ok(())`, since I'm not sure what I should be returning from this function?

[the remark this answers is not in the document]
Comment (brendan-oconnell) [reply thoth-api/src/graphql/model.rs]: Refactored this significantly, so I'm going to resolve this comment and you can take a look at the new version.

File: thoth-app/src/models/location/location_query.rs (hunk 2 above)
Comment (brendan-oconnell) [reply thoth-app/src/models/location/location_query.rs:1]: Used on lines 34 and 40: `#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]`

Comment (rhigman) [reply thoth-app/src/models/location/location_query.rs:1]: Sorry, I meant the location_query module itself.

File: thoth-app/src/models/mod.rs (hunk 3 above)
Comment (brendan-oconnell) [reply thoth-app/src/models/mod.rs:20]: Because everything in the view is coming from `Publication`, rather than `Location`, these are never used for `Location` However, `FetchLocation` and  `FetchActionLocation` have to stay in `thoth-app/src/models/location/location_query.rs` to conform to the structure in `macro_rules! graphql_query_builder`. 

So if I don't use `#[allow(dead_code)]`, I get an error from `cargo clippy --all --all-targets --all-features -- -D warnings`:

```
error: type alias `FetchLocation` is never used
    --> thoth-app/src/models/location/location_query.rs:30:5
```

Hope this makes sense!

File: thoth-app/src/models/location/location_query.rs (hunk 2 above)
Comment (brendan-oconnell) [reply thoth-app/src/models/location/location_query.rs]: Yes, you're right... I think this was leftover from an earlier iteration, before everything was going through `Publication`. I've deleted it.

File: thoth-app/src/models/location/update_location_mutation.rs (hunk 1 above)
Comment (rhigman) [reply thoth-app/src/models/location/update_location_mutation.rs:24]: As discussed, I retract this comment! :upside_down_face:

File: thoth-app/src/models/mod.rs (hunk 3 above)
Comment (rhigman) [reply thoth-app/src/models/mod.rs:20]: Is this still needed now that the `location_query` module has been removed?

File: thoth-api/src/model/location/crud.rs [hunk 8]
```diff
@@ -126,7 +126,101 @@ impl Crud for Location {
...
+    ) -> ThothResult<Self> {
+        // if changes to a location don't change its canonical or non-canonical status, only update that location.
+        if data.canonical == self.canonical {
+            let mut connection = db.get()?;
+            connection.transaction(|connection| {
+                match diesel::update(location::table.find(&self.location_id))
+                    .set(data)
+                    .get_result::<Self>(connection)
+                {
+                    // On success, create a new history table entry.
```
Comment (rhigman) [inline thoth-api/src/model/location/crud.rs:168]: To avoid repeating this logic across two branches, could store the `diesel::update` result(s) to a variable and then `match` on it at the very end of the function.

File: thoth-api/src/model/location/crud.rs [hunk 9]
```diff
@@ -126,7 +126,101 @@ impl Crud for Location {
...
+
+    fn update(
+        &self,
+        db: &crate::db::PgPool,
+        data: &PatchLocation,
+        account_id: &Uuid,
+    ) -> ThothResult<Self> {
+        // if changes to a location don't change its canonical or non-canonical status, only update that location.
+        if data.canonical == self.canonical {
+            let mut connection = db.get()?;
```
Comment (rhigman) [inline thoth-api/src/model/location/crud.rs:162]: Could do this outside the branches instead of repeating it both here and in `else`

File: thoth-app/src/models/mod.rs (hunk 3 above)
Comment (brendan-oconnell) [reply thoth-app/src/models/mod.rs:20]: no, it doesn't seem they are -- good catch!

File: thoth-api/src/model/location/crud.rs (hunk 8 above)
Comment (brendan-oconnell) [reply thoth-api/src/model/location/crud.rs]: great idea, thanks!

**Review by ja573 — CHANGES_REQUESTED** [review]
This is all working perfectly. Just have a few comments about streamlining certain bits (which I realise you're just inheriting! So feel free to correct them in the original)

File: thoth-api/src/model/location/crud.rs [hunk 10]
```diff
@@ -126,7 +126,98 @@ impl Crud for Location {
         crate::model::publication::Publication::from_id(db, &self.publication_id)?.publisher_id(db)
     }
 
-    crud_methods!(location::table, location::dsl::location);
+    // `crud_methods!` cannot be used for update(), because we need to execute multiple statements
+    // in the same transaction for changing a non-canonical location to canonical.
+    // These functions recreate the `crud_methods!` logic.
+    fn from_id(db: &crate::db::PgPool, entity_id: &Uuid) -> ThothResult<Self> {
+        let mut connection = db.get()?;
+        match location::table
```
[the remark this answers is not in the document]
Comment (brendan-oconnell) [reply thoth-api/src/model/location/crud.rs:134]: Great, I've refactored in work_relation/crud.rs and as many as I could do easily in mod.rs!

File: thoth-api/src/model/location/crud.rs [hunk 11]
```diff
@@ -126,7 +126,98 @@ impl Crud for Location {
...
+
+        connection.transaction(|connection| {
+            diesel::insert_into(location::table)
+                .values(data)
+                .get_result::<Self>(connection)
+                .map_err(|e| e.into())
+        })
+    }
+
+    fn update(
```
[the remark this answers is not in the document]
Comment (brendan-oconnell) [reply thoth-api/src/model/location/crud.rs:154]: Hey, this is all great, and thanks for the streamlining! It appears that we still need to clone `old_canonical_location.location_id`; when I remove `clone()`, I get a compiler error, "cannot move out of `old_canonical_location` because it is borrowed". So I put the `clone()` back in.

[the remark this answers is not in the document]
Comment (brendan-oconnell) [reply thoth-api/src/model/location/crud.rs:154]: Disregard - I got it to work without using `clone()`! Not sure what was going on.

**Review by ja573 — APPROVED** [review]
